### ISSUE TITLE
Why can not compile Example1.Extension?

### ISSUE THREAD

**Issue by CreateAndInject** [0xd4d/dnSpy#260]
using Microsoft.VisualStudio.Text;

All of its references don't contains this namespace.

**Comment by 0xd4d** [comment]
Did you read the wiki: https://github.com/0xd4d/dnSpy/wiki/Building-dnSpy

**Comment by CreateAndInject** [comment]
I don't use git, I just download the source code.

**Comment by TiberiumPY** [comment]
@CreateAndInject then you need manually download source code from submodules

**Comment by CreateAndInject** [comment]
I don't compile the whole solution, just Example1.Extension project.

I can compile it with these refercences:
Microsoft.VisualStudio.Text.Data
Microsoft.VisualStudio.Text.Logic
Microsoft.VisualStudio.Text.UI

But I want to know: why the Example1.Extension.csproj don't contains these?
**So, everybody should add these refercences manually?**

  <ItemGroup>
    <Reference Include="PresentationCore" />
    <Reference Include="PresentationFramework" />
    <Reference Include="System" />
    <Reference Include="System.ComponentModel.Composition" />
    <Reference Include="System.Core" />
    <Reference Include="System.Xaml" />
    <Reference Include="WindowsBase" />
  </ItemGroup>

 <ItemGroup>
    <ProjectReference Include="..\..\..\dnSpy\dnSpy.Contracts.DnSpy\dnSpy.Contracts.DnSpy.csproj">
      <Project>{cb6f6372-9479-41e2-83a8-07854326bca8}</Project>
      <Name>dnSpy.Contracts.DnSpy</Name>
    </ProjectReference>
    <ProjectReference Include="..\..\..\dnSpy\dnSpy.Contracts.Logic\dnSpy.Contracts.Logic.csproj">
      <Project>{bb748e78-376f-429a-8d83-9a8946ec179f}</Project>
      <Name>dnSpy.Contracts.Logic</Name>
    </ProjectReference>
    <ProjectReference Include="..\..\..\Libraries\dnlib\src\dnlib.csproj">
      <Project>{fdfc1237-143f-4919-8318-4926901f4639}</Project>
      <Name>dnlib</Name>
    </ProjectReference>
  </ItemGroup>

**Comment by 0xd4d** [comment]
Nuget references are also in `project.json`. Your Visual Studio should already restore them before the build. If that doesn't happen, compile once, restart it, try again. If it still doesn't work make sure you're using VS2015.

**Comment by CreateAndInject** [comment]
.Net framework has already contains these references, why use Nuget references?
It means we must connect network to build it.
Even if connecting network, may can not download these becuase of Great Firewall of China.

**Comment by UlyssesWu** [comment]
@CreateAndInject You are asking why we use nuget. That's not even a question about dnspy. I think you can find it out by yourself, through google or baidu etc.
You should know people in .NET world all use nuget. Nuget and nuget.org is not blocked by GFW (sometimes it's slow though, but since you are here at github...). 
Even if you cannot download that dependencies in nuget (Microsoft.VisualStudio.Text.UI.Wpf.dll etc.), you can find them in the latest binary and then reference them.

**Comment by CreateAndInject** [comment]
But sometimes I don't know which dlls, VisualStudio only tell you can not find some namespaces and types, but I don't know which dll contains these namespaces/types.

**Comment by CreateAndInject** [comment]
@UlyssesWu  "You should know people in .NET world all use nuget."
That's because:
1. Those softwares must connect network when running, but dnSpy not.
2.Those references are not part of .net framework, but these dlls are. So, why we should get them from network when the .net framework has already contains them?

**Comment by UlyssesWu** [comment]
@CreateAndInject First question, open project.json and you can see "Microsoft.VisualStudio.Text.UI.Wpf".
Second one: "Those softwares must connect network when running."  That's not true at all. What could I say? ignorant?
"Those references are not part of .net framework, but these dlls are. " You are misunderstanded. See project.json again. Nuget only get one package for you --- "Microsoft.VisualStudio.Text.UI.Wpf". Not those in .csproj. 
“你们有一个好，github上跑到什么地方，你们比其他的西方程序员啊，跑得还快。但是呢，问来问去的问题啊，都 too simple，sometimes naïve！” ——希望您在问问题之前，努力提高自己的姿势水平，以及英语水平。

**Comment by 0xd4d** [comment]
.NET Framework doesn't contain all required references so nuget is needed to get these assemblies.

I don't see a problem, you already have to connect to the Internet to download the source code, and then you can download the nuget packages. They're cached in `%USERPROFILE%\.nuget` so you don't need to re-download them every time the code is compiled.

**Comment by CreateAndInject** [comment]
@UlyssesWu “你们有一个好，github上跑到什么地方，你们比其他的西方程序员啊，跑得还快。但是呢，问来问去的问题啊，都 too simple，sometimes naïve！”

What's this? Why does it contain so many grammar errors, I don't understand.

**Comment by CreateAndInject** [comment]
@0xd4d 
1. Microsoft.VisualStudio.Text.Data,Microsoft.VisualStudio.Text.Logic,Microsoft.VisualStudio.Text.UI
All the 3 dlls are in the GAC. So, can I say these 3 are the part of .net framework? If that, why use nuget in the project "Example1.Extension"?

2. To compile "Example1.Extension", we need these 3 dlls, and don't need :Microsoft.VisualStudio.Text.UI.Wpf:

So why project.json contains "Microsoft.VisualStudio.Text.UI.Wpf" not these 3 dlls?

**Comment by 0xd4d** [comment]
Just because you have the files doesn't mean everyone have them. They're not part of the .NET Framework.

**Comment by CreateAndInject** [comment]
How to build this solution?
I download the source and all submodule source, than use F6 in Visual Studio, but these are too many errors.

**Comment by 0xd4d** [comment]
It's because you don't check out the correct source codes.

Follow the Wiki and it'll work. Otherwise you're on your own.

**Comment by CreateAndInject** [comment]
I have already tested several times, it seems that Visual Studio 2015 RTM without any updates can not compiled this solusion(circular reference/An item with the same key has been added/others), but Visual Studio 2015 update 3 can.(without any other operation , press F6 directly)
